### PR TITLE
Add travis build

### PR DESCRIPTION
**Pull request by erikd**
Only builds with GHC 7.10 and 8.0 for now. Can't build in Travis with 7.8.4 because the compiler available hasn't been built with profiling and hence the tests fail. Building with 8.2 is also busted and will be fixed in an upcoming PR.

@haskell-mafia

### CODE REVIEW

**Comment by erikd** [comment]
Ok, I finally have something that builds with ghc 7.8.4, 7.10.3 and 8.0.2.

Unfortunately the builds sometimes fail when building under travis. This can probably fixed by simplifying some of the tests.

File: src/Mafia/Script.hs [hunk 1]
```diff
@@ -136,7 +136,7 @@ parseScript path source =
 
 renderSubmoduleUrl :: Submodule -> Text
 renderSubmoduleUrl s =
-  "git@github.com:" <> submoduleUser s <> "/" <> submoduleRepo s
+  "https://github.com/" <> submoduleUser s <> "/" <> submoduleRepo s
```
Comment (nhibberd) [inline src/Mafia/Script.hs]: Isn't this a breaking change for mafia scripts? can the protocol be added to the Pragma in a compatible way? 

e.g.

```
data Submodule =
  Submodule {
      submoduleProtocol :: !Protocol
    , submoduleUser :: !Text
    , submoduleRepo :: !Text
    , submoduleCommit :: !(Maybe Text)
    } deriving (Eq, Ord, Show)

Current implicit behaviour:
{-# SUBMODULE ambiata/p@xxx #-} -> Submodule Ssh "ambiata" "p" "xxx"

Explicit protocol:
{-# SUBMODULE ssh ambiata/p@xxx #-} -> Submodule Ssh "ambiata" "p" "xxx"

{-# SUBMODULE https ambiata/p@xxx #-} -> Submodule Https "ambiata" "p" "xxx"
```

I am sure there is a better way of describing the protocols, just first thing that came to mind that would keep compatibility

Comment (erikd) [reply src/Mafia/Script.hs]: Yep, that makes sense. Will look into it.

Comment (erikd) [reply src/Mafia/Script.hs]: Actually, why not make the protocol and the server explicit like:
```
{-# SUBMODULE git@github.com ambiata/p@xxx #-}
{-# SUBMODULE https://gitlab.com ambiata/x@xxx #-}
```
so that other services like Gitlab and Bitbucket can be used.

Its harder to make this backwards compatible though.

**Comment by HuwCampbell** [comment]
Just throwing in something slightly orthogonal.

I've wanted to explicitly specify an sdist tarball which isn't on hackage as a submodule. So extensions might be interesting to consider too.

**Comment by erikd** [comment]
@HuwCampbell By all mean raise an issue, but this PR is about getting mafia building in travis without breaking any current functionality.

**Comment by erikd** [comment]
New version allows:

```
{-# SUBMODULE ambiata/p@xxx #-}      
{-# SUBMODULE git@github.com ambiata/p@xxx #-}  
{-# SUBMODULE https://gitlab.com ambiata/x@xxx #-}
```
where the first assumes the location is `git@github.com` and is therefore backwards compatible with what we had.

Also added a property test to round-trip those `SUBMODULE` lines.

File: src/Mafia/Script.hs [hunk 2]
```diff
@@ -5,11 +5,20 @@ module Mafia.Script (
     ScriptError(..)
   , renderScriptError
   , runScript
+
+  -- For testing.
+  , Pragma (..)
```
Comment (charleso) [inline src/Mafia/Script.hs:10]: This whole script change seems unrelated to building in travis. Should it be it's own PR?

File: src/Mafia/Script.hs (hunk 1 above)
Comment (charleso) [reply src/Mafia/Script.hs]: > Actually, why not make the protocol and the server explicit like:

I'm actually 100% on board with this, I never liked hard coding to github. That said, my suggestion is that you should also fall back to the previous behaviour to remain backwards compatible (as well as keeping with Jacob's original intention)

**Comment by charleso** [comment]
I am curious why not keep the original ambiata submodules for disorder/x/twine rather than forking them? Especially if the plan is to just inline them at some point. Are the forks different?

**Comment by charleso** [comment]
I'm 👍 for a travis build, but the script change seems like it should be kept separate _and_ remain backwards compatible.

**Comment by erikd** [comment]
I forked the ambiata submodules because as they were, they would not clone over a HTTPS transport, but required SSH transport.

The idea is to get this building in travis and then remove the submodules so that this can then be put in Hackage.

**Comment by charleso** [comment]
> I forked the ambiata submodules because as they were, they would not clone over a HTTPS transport, but required SSH transport.

"required ssh"? I'm curious why not - what stops it being cloned (that's news to me)?

**Comment by charleso** [comment]
> what stops it being cloned (that's news to me)?

Ahh sorry you mean their submodules are ssh and not http? I'm with you now. Sorry about that.

**Comment by erikd** [comment]
Yeah, the script tests needed changes because they *required* SSH access to clone the required repos.

I'm pretty sure this is the absolute minimum set of changes to get travis builds working :).

Edit: Oh, and the Script change is backwards compatible :)

**Review by tmcgilchrist — APPROVED** [review]
Getting this building on travis is a good first step. 
We'll inline / remove the submodules in later PRs

**Comment by erikd** [comment]
@nhibberd @charleso Ok with this? You'll have to acknowldge via an actual "review".
